pep-0546.txt

```diff
@@ -0,0 +1,118 @@
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 30-May-2017
+
+
+Abstract
+========
+
+Backport ssl.MemoryBIO and ssl.SSLObject classes from Python 3 to Python
+2.7 to enhance the overall security of Python 2.7.
+
+
+Rationale
+=========
+
+While Python 2.7 is getting closer to its end-of-line (scheduled for
+2020), it is still used on production and the Python community is still
+responsible for its security.
+
+PEP 543
+-------
+
+The :pep:`543` defines a new TLS API for Python which would enhance the
+Python security: give access to the root certificate authorities on
+Windows and macOS by using native APIs, instead of OpenSSL. A side effect
+is that it gives access to certificates installed locally by system
+administrators, allowing to use "company certificates" without having to
+modify each Python application and so validate correctly TLS
+certificates (instead of having to ignore or bypass the TLS certificate
+validation).
+
+For practical reasons, Cory Benfield would like to first implement an
+I/O-less class similar to ssl.MemoryBIO and ssl.SSLObject for the
+:pep:`543`, and provide a second class based on the first one to use
+sockets or file descriptors.  This design would help to factorize the
+code to support more backends and simplify testing and auditing. Later,
+optimized classes using directly sockets or file descriptors may be
+added for performance.
+
+While the :pep:`543` defines an API, the PEP would only make sense if it
+comes with at least one complete and good implementation. The first
+implementation will be based on the ``ssl`` module of the Python
+standard library.
+
+In a perfect world, all applications would already run on Python 3 since
+Python 3.0 was released. In practice, many applications still run on
+production on top of Python 2.7. To make the new TLS API more widely
+used, it should be usable on all Python versions currently supported:
+Python 2.7, 3.5, 3.6. Otherwise, some applications would have to wait
+until they drop Python 2 support to be able to use the new TLS API.
+
+requests and pip
+----------------
+
+There are plans afoot to look at moving Requests to a more event-loop-y
+model, and doing so basically mandates a MemoryBIO. In the absence of a
+Python 2.7 backport, Requests is required to basically use the same
+solution that Twisted currently does: namely, a mandatory dependency on
+`pyOpenSSL <https://pypi.python.org/pypi/pyOpenSSL>`_.
+
+The `pip <https://pip.pypa.io/>`_ program has to embed all its
+dependencies for pratical reason. Since pip depends on requets, it means
+that it would have to embed a copy of pyOpenSSL. That would imply
+usability pain to install pip. Currently, pip doesn't support embedding
+C extensions which must be compiled on each platform and so require a C
+compiler.
+
+
+Changes
+=======
+
+Add ``MemoryBIO`` and ``SSLObject`` classes to the ``ssl`` module of
+Python 2.7.
+
+The code will be backported and adapted from the master branch
+(Python 3).
```

Comment:
I expect the code to be exactly the same. I don't think that MemoryBIO or SSLObject changed much since Python 3.5.

I prefer to backport from master to ease comparison of 2.7 and master branches, to easy re-sync later. As explained in another paragraph: reduce the diff between these two branches.